components/style/properties/properties.mako.rs

```diff
@@ -1504,6 +1297,180 @@ impl PropertyDeclaration {
+                            },
+                            Err(()) => Err(PropertyDeclarationParseError::InvalidValue),
+                        }
+                    % else:
+                        Err(PropertyDeclarationParseError::UnknownProperty)
+                    % endif
+                }
+            % endfor
+            },
+            PropertyId::Shorthand(id) => match id {
+            % for shorthand in data.shorthands:
+                ShorthandId::${shorthand.camel_case} => {
+                    % if not shorthand.allowed_in_keyframe_block:
+                        if rule_type == CssRuleType::Keyframe {
+                            return Err(PropertyDeclarationParseError::AnimationPropertyInKeyframeBlock)
+                        }
+                    % endif
+                    % if shorthand.internal:
+                        if context.stylesheet_origin != Origin::UserAgent {
+                            return Err(PropertyDeclarationParseError::UnknownProperty)
+                        }
+                    % endif
+                    % if not shorthand.allowed_in_page_rule:
+                        if rule_type == CssRuleType::Page {
+                            return Err(PropertyDeclarationParseError::NotAllowedInPageRule)
+                        }
+                    % endif
+
+                    ${property_pref_check(shorthand)}
+
+                    match input.try(|i| CSSWideKeyword::parse(context, i)) {
+                        Ok(keyword) => {
+                            % if shorthand.name == "all":
+                                declarations.all_shorthand = AllShorthand::CSSWideKeyword(keyword);
+                            % else:
+                                % for sub_property in shorthand.sub_properties:
+                                    declarations.push(PropertyDeclaration::CSSWideKeyword(
+                                        LonghandId::${sub_property.camel_case},
+                                        keyword,
+                                    ));
+                                % endfor
+                            % endif
+                            Ok(())
+                        },
+                        Err(()) => {
+                            shorthands::${shorthand.ident}::parse_into(declarations, context, input)
+                                .map_err(|()| PropertyDeclarationParseError::InvalidValue)
+                        }
+                    }
+                }
+            % endfor
+            }
+        }
+    }
+}
+
+const MAX_SUB_PROPERTIES_PER_SHORTHAND_EXCEPT_ALL: usize =
+    ${max(len(s.sub_properties) for s in data.shorthands if s.name != "all")};
+
+type SourcePropertyDeclarationArray =
+    [PropertyDeclaration; MAX_SUB_PROPERTIES_PER_SHORTHAND_EXCEPT_ALL];
+
+/// A stack-allocated vector of `PropertyDeclaration`
+/// large enough to parse one CSS `key: value` declaration.
+/// (Shorthands expand to multiple `PropertyDeclaration`s.)
```

Comment:
It'd be nice to elaborate on how it's used here, for example, saying that it's used as a scratch buffer for each `key: value` in a CSS declaration block.